[FLOC-3770] Add MongoDB backend

zubron opened the pull request:
This change adds support for using a MongoDB backend. It uses [TxMongo](https://github.com/twisted/txmongo) as the driver.

The same set of tests that are available for the InMemoryBackend are used with this backend. If the MongoDB backend is selected, then the default connection details are `127.0.0.1:27017`. This can be changed using the newly added command line switches.

The documentation on how to use MongoDB is pending however it would be helpful to have the code changes reviewed so far.

avg-I committed:
add disconnect method to the backend interface

And arrange for that method to be called at appropriate times.

The main script now has two services, an HTTP endpoint service and
a basic BackendService.  These services are managed under a top-level
MultiService.

avg-I committed:
a nit: use correct order for expected and actual test results

avg-I committed:
add a generic handler for failures without a specific handler

This handler puts error description into the JSON format and sets
the response code to INTERNAL_SERVER_ERROR.
This is mostly designed for unexpected exceptions arising from a backend
as any expected (documented) failures should have specific handlers.

avg-I committed:
add a "real" backend that uses txmongo driver to work with MongoDB

Note the following limitations:
- txmongo does not support 'sort' parameter for collection.find() method
  and the parameter is silently ignored, so query() method of the backend
  is broken now
- currently the backend relies on auto-generated IDs being monotonically
  increasing, but that's no guaranteed
- moreover, we should use a timestamp embedded into documents for sorting

Also, it's not clear if we should include the "_id" field with the
retrieved documents.  If we decide to do so, then we need to take care
to properly serialize its ObjectId value (as a string, probably).
Also, it would make sense to return the full document including the _id
field as a response to a POST that creates a new document.

Note that at the moment the new backend is not used and thus not tested.

avg-I committed:
split test class into mixin with actual tests and backend-specific class

avg-I committed:
use relative import

avg-I committed:
add functional tests for TxMongoBackend based on BenchmarkAPITestsMixin

avg-I committed:
use Travis CI only with tests that do not have environmental dependencies

In other words, do not run functional tests.

avg-I committed:
Merge branch 'FLOC-3302-query-api' into FLOC-3770-txmongo-backend

avg-I committed:
update TxMongoBackend to sort results by 'timestamp' field

avg-I committed:
Merge branch 'FLOC-3302-query-api' into FLOC-3770-txmongo-backend

avg-I committed:
mongodb: remove only test entries instead of clearing the whole collection

avg-I committed:
Merge branch 'FLOC-3302-query-api' into FLOC-3770-txmongo-backend

avg-I committed:
Merge branch 'FLOC-3302-query-api' into FLOC-3770-txmongo-backend

zubron committed:
Merge remote-tracking branch 'origin/master' into FLOC-3770-txmongo-backend

zubron committed:
Only perform clean up if a document was created

zubron committed:
Ensure the TxMongo backend behaviour matches the InMemory backend.

MongoDB will return all documents for a limit of 0, however this
change will prevent this behaviour and will instead return 0 documents.

zubron committed:
Run TxMongo backend tests on Travis

zubron committed:
Use more recent version of MongoDB

zubron committed:
Add options to specify backend and connection details

zubron committed:
Exclude fields from results rather than deleting afterwards

zubron committed:
Add space in travis configuration file

james-w commented on benchmark/functional/test_httpapi.py:
```diff
@@ -0,0 +1,28 @@
+        def add_cleanup(response):
+            if response.code == CREATED:
+                location = response.headers.getRawHeaders(b'Location')[0]
+                self.addCleanup(lambda: self.agent.request("DELETE", location))
```
:question: Did you consider a "DELETE all" function that could be done at the start
of each mongo test? The performance would be different, but it would likely
mean fewer cascading test failures if something goes wrong such that this
doesn't run.

avg-I replied:
Actually that's how I originally implemented it, but then I decided that it was safer to remove individual records just in case the tests would be run against a wrong MongoDB server by accident, or some such.
By I do not have a strong opinion.

jml replied:
:question: Perhaps this should be part of the base class?

avg-I replied:
It certainly can be as now it does not do anything backend-specific (previously it did).
So, I'll move the code.

james-w commented on benchmark/httpapi.py:
```diff
@@ -251,20 +368,59 @@ def create_api_service(endpoint):
 
     optParameters = [
         ['port', None, 8888, "The port to listen on", int],
+        ['backend', None, 'in-memory', "The persistence backend to use", str],
```
:question: would it be possible to list the supported backends here without duplicating the list?

avg-I replied:
Yes, it's possible. I am making the change.

james-w commented on benchmark/httpapi.py:
```diff
@@ -277,18 +433,23 @@ def main(reactor, args):
         raise SystemExit(1)
 
+    def get_backend_from_options(options):
+        backend = _BACKENDS[options['backend']]
```
:question: Something nicer than a KeyError on a typo would be nice, but incredibly low priority.

avg-I replied:
I am doing this anyway.

james-w commented on benchmark/httpapi.py:
```diff
@@ -277,18 +433,23 @@ def main(reactor, args):
-    service = create_api_service(TCP4ServerEndpoint(reactor, options['port']))
 
-    # XXX Make startService() raise an exception on an error
-    # instead of just logging and dropping it.
```
:question: I realise this comment isn't added by you, but it's not really clear to me
how we fix this. We don't seem to override that method in this codebase,
so is it referring to a change in Twisted.

If it's our code then moving it to the code that we should change would be
better. If it's twisted code then mentioning that, maybe with a bug reference
would be ideal.

If you aren't sure either then that's fine. If no-one is sure maybe we should
just delete the comment.

avg-I replied:
Actually the comment is about the next line of the code, about something we do already but in a somewhat ugly way:
```.py
    service._raiseSynchronously = True
```
Probably the comment could use some improvement to make that clearer.

james-w commented on benchmark/httpapi.py:
```diff
@@ -100,6 +112,104 @@ def delete(self, id):
+        try:
+            object_id = ObjectId(id)
+        except InvalidId:
+            raise ResultNotFound()
```
If it's an InvalidId then it's probably a programming error as we shouldn't
be called with an Id that we didn't generate? I think that letting the InvalidId
exception propagate would avoid masking that error.

avg-I replied:
Actually the ID comes from an HTTP request and the common HTTP handling code can not know if a string is a valid ID for a particular backend.  In other words, the ID is used in external exchanges, not internally.

james-w commented on benchmark/httpapi.py:
```diff
@@ -100,6 +112,104 @@ def delete(self, id):
+
+        def post_process(result):
+            if result is None:
+                raise ResultNotFound()
```
Could this include some more information about the request for debugging
purposes? At least the id that was requested would be good.

avg-I replied:
Okay.

james-w commented on benchmark/httpapi.py:
```diff
@@ -100,6 +112,104 @@ def delete(self, id):
+        try:
+            object_id = ObjectId(id)
+        except InvalidId:
+            raise ResultNotFound()
```
Similar comments to above about InvalidId and a more informative ResultNotFound.

james-w commented:
Hi,

Thanks for the branch, it was very clear to read, with some good comments.

I've left a few comments, but most are low priority, or me just asking to learn.

There are no direct tests of the mongo backend, but I think that's reasonable
because it's a pretty thin layer over txmongo. However, it may warrant a few tests
to check the error handling.

I'm happy for this to land once you have made any changes you wish to make,
however it will need a mentor review first. I will ask @jml, but if you find someone
else that is fine too.

Thanks,

James

jml commented on benchmark/httpapi.py:
```diff
@@ -100,6 +112,104 @@ def delete(self, id):
+    """
+    def __init__(self, hostname="127.0.0.1", port=27017):
+        connection = MongoConnectionPool(host=hostname, port=port)
+        self.collection = connection.benchmark.results
```
Does this actually make a connection? If so, then it would perhaps be better to move the connection logic to a method, and leave the constructor being super boring.

avg-I replied:
It creates a connection object, a database object and a collection object.
But no actual connection is established at that point (i.e. no network packets are sent), that is done behind the scenes when the collection object gets used.

jml commented on benchmark/httpapi.py:
```diff
@@ -251,20 +368,59 @@ def create_api_service(endpoint):
+    backend_service.setServiceParent(top_service)
+
+    # XXX Make startService() raise an exception on an error
+    # instead of just logging and dropping it.
```
I don't know if this comments means that the code does not do this but we should change it to do so in the future or whether it describes what it does. If the latter, please remove the 'XXX'.

avg-I replied:
The reason I put XXX there is that the code sets a "private" field to get the desired effect, so that is not quite nice.
Maybe that's not good enough reason?

jml replied:
It's an OK reason, but maybe make it more clear. e.g.

> XXX: Setting _raiseSynchronously makes startService raise an exception on error rather than just logging and dropping. This should be a public API, $TWISTED_BUG_HERE

avg-I replied:
Okay.

jml commented:
Mentor review approval

avg-I committed:
move the test cleanup code to the base class

The code no longer does anything backend-specific.

avg-I committed:
improve parsing and displaying of server options

- display a list of supported backends
- produce a nice message for an unknown backend

avg-I committed:
first argument of txmongo's find_one() is different from 'filter' argument

It seems that the first argument that actually specifies the query is
named 'spec' in txmongo and the 'filter' argument does something different
(not sure what).

Previously the code passed the tests because the testing was done
essentially against an empty collection, so there was no chance of
acting on a wrong record.  Now the tests have been updated to have
an unrelated record.

avg-I committed:
improve reporting of missing and invalid result identifiers

- differentiate between missing results and bad IDs
- include the requested ID with an exception
- provide useful messages about the problems to a client

avg-I committed:
clarify a code comment
